### PR TITLE
Replace the container launcher with feature detection, update launcher handling

### PR DESCRIPTION
**Pull request by djmitche**
### What does this PR do?

This contains some house-keeping changes, isolating launchers somewhat from the rest of the logs agent, followed by a replacement of the container launcher with feature detection.

Previously, the container launcher didn't launch anything, but just tried to start up either the docker or kubernetes launchers, checking both features and whether kubelet or the docker API were available.  It took whichever appeared first, preferring one over the other based on the `logs_config.k8s_container_use_file` config value.

This will soon become problematic, as those launchers are both doing more than they should, and will be broken into components.  So, we'll need a way for multiple components to determine whether they should start up, which is what this PR does.

### Describe how to test/QA your changes

Verify that container_collect_all detects containers in various configurations, either via docker or via kubernetes.  See https://github.com/DataDog/datadog-agent/pull/8148 for background.

Expected results:
 - if neither kubelet nor docker can start, containers are not detected
 - if kubelet can start but docker cannot, containers are detected using kubelet
 - if docker can start but kubelet cannot, containers are detected using docker
 - if both can start
     - with DD_LOGS_CONFIG_K8S_CONTAINER_USE_FILE=false, containers are detected using docker
     - with DD_LOGS_CONFIG_K8S_CONTAINER_USE_FILE=true, containers are detected using kubectl

I tested this on a dedicated GKE cluster, with the agent installed per https://docs.datadoghq.com/agent/kubernetes/?tab=daemonset.  To disable kubelet, I set 
```
            - name: DD_KUBELET_TLS_VERIFY
              value: "true"
            - name: DD_KUBELET_CLIENT_CA
              value: "/tmp/nosuch"
```
to disable Docker, I commented out
```
            - name: runtimesocketdir
              mountPath: /host/var/run
              mountPropagation: None
              readOnly: true
```

Both launchers start up with messages like
```
2022-03-11 20:51:58 UTC | CORE | INFO | (pkg/logs/internal/launchers/kubernetes/launcher.go:84 in Start) | Starting Kubernetes launcher
```

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Comment by djmitche** [comment]
Ugh, looks like there's a lot more mess here than I thought -- the unit tests appear to be sensitive to which of docker and kubernetes are available..

**Comment by djmitche** [comment]
This is still ready for review, but I'll work on the QA steps directly in parallel.

**Comment by djmitche** [comment]
It occurred to me that I can do one-better here, avoiding some races.

It's already the case in this PR that only one of the docker and k8s launchers starts, based on feature detection and configuration.

Once that's done, the current implementation waits until kubelet or docker is available before subscribing to integration.Configs from AD.  However, that leads to a race where, if AD finds kubelect/docker first, it may send a bunch of Configs before the launcher starts listening.  But doing the opposite may lead to AD getting blocked sending to a channel when the launcher is not yet reading from it.

What I realized is that the k8s launcher already has the ability to retry processing integration.Config's, and it'd be easy to add that to the docker launcher as well.  And, I can use that support to also retry Config's when kubelet/docker aren't available.

**Comment by djmitche** [comment]
OK, I think I'm happier with this.  The retries in the last commit are a bit of a bodge, but
 * They're unlikely to ever be used (how would AD tell logs-agent about a container, if there's no container runtime??)
 * They're mostly used for container_collect_all, which will move into AD soon

**Comment by djmitche** [comment]
This represents a _bit_ of a behavior change.  The old behavior was:

 - if neither kubelet nor docker can start, neither launcher runs
 - if kubelet can start but docker cannot, the kubernetes launcher runs
 - if docker can start but kubelet cannot, docker launcher runs
 - if both can start
     - with DD_LOGS_CONFIG_K8S_CONTAINER_USE_FILE=false, the docker launcher runs
     - with DD_LOGS_CONFIG_K8S_CONTAINER_USE_FILE=true, the kubernetes launcher runs

The new behavior is:

 - if DD_LOGS_CONFIG_K8S_CONTAINER_USE_FILE=false, the docker launcher runs
    - if the docker socket is not available, nothing happens
 - if DD_LOGS_CONFIG_K8S_CONTAINER_USE_FILE=true, the kubernetes launcher runs
    - if kubelet is not available, nothing happens

this is because whenever running in kubernetes, both the docker and kubernetes features are enabled, and DD_LOGS_CONFIG_K8S_CONTAINER_USE_FILE is what defines the preference between those.  Basically, I've changed the "have a preference but use whatever starts working first" to "have a preference".

**Comment by olivielpeau** [comment]
This behavior change may have a significant functional impact.

One example I can think of is that, since `DD_LOGS_CONFIG_K8S_CONTAINER_USE_FILE=false` is the default value of this config option including on k8s environments where the docker socket may not be available (or where docker isn't used at all), on such environments I would expect log collection to stop working by default with this change. (Is that correct?)

Also, does this PR change the behavior implemented in https://github.com/DataDog/datadog-agent/pull/8148 ? (that attempts retrying both launchers until one succeeds)

More generally, we'd like the logs agent launcher logic to be resilient to a temporary unavailability of the docker or the kubernetes launchers at startup (I don't know if this PR changes anything about that).

Sorry for the unstructured thoughts here. @gh123man has directly worked on this logic in the past and may be able to provide more structured thoughts.

File: pkg/logs/internal/util/container_or_k8s.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	k := config.IsFeaturePresent(config.Kubernetes)
+
+	switch {
+	case c && !k:
+		return LogContainers
+	case k && !c:
+		return LogPods
+	case k && c:
+		// prefer kubernetes if k8s_container_use_file is set
+		if config.Datadog.GetBool("logs_config.k8s_container_use_file") {
```
Comment (gh123man) [inline pkg/logs/internal/util/container_or_k8s.go:41]: On @olivielpeau 's comment. I think the change in behavior here is that previously we relied on both the feature being enabled AND the util (docker or K8s) being available where the util would retry. 

IIRC the core case https://github.com/DataDog/datadog-agent/pull/8148 was trying to solve was that both docker an K8s were available as features. But the agent would only (forever) retry docker if the kublet failed to connect once. 

With this PR - I think the agent will make the correct decision based on the config flags and feature availability, but will not dynamically failover if one of the utils is unavailable. I think the bigger question is - what scenarios rely specifically on the failover behavior? I think in a perfect world an agent configuration should decide what features it expects to use explicitly. But I think we should avoid the possibility of breaking existing agents.

**Comment by djmitche** [comment]
For reference, "Docker" in this context means Docker, Containerd, Cri, or Podman.
The features are currently defined as follows:

 * _Kubernetes_ - either $KUBERNETES_SERVICE_PORT or $KUBERNETES is set
 * _Docker_ - either $DOCKER_HOST is set or one of the docker sockets exists and is readable (but notably, there is no check whether the docker daemon is actually listening there)
 * _Containerd_ / _Cri_ -
   * if `cri_socket_path` is set; or if `/.dockerenv` doesn't exist and one of the default CRI sockets exists and is readable (but notably, there is no check whether the CRI daemon is actually listening there); then
   * if either $KUBERNETES_SERVICE_PORT or $KUBERNETES is set, then feature _Cri_ is enabled
   * if `containerd` appears in the detected CRI socket filename, then feature _Containerd_ is enabled.
 * _Podman_ - if any of the podman socket paths exist (without testing readability)

It may be that DD_LOGS_CONFIG_K8S_CONTAINER_USE_FILE was defined in such an implementation-specific way that it's impossible to make this change and fix the race condition mentioned above.  Maybe this isn't worth changing at all?

**Comment by djmitche** [comment]
Closed in favor of #11289.

**Pull request closed without merging**